### PR TITLE
GHA/windows: enable MulitSSL in an MSVC job

### PR DESCRIPTION
**Pull request by talregev**
Ref: 98da147b18fa4963efb425df8400ca40e7526646 #14305
Closes #14276

### CODE REVIEW

**Comment by talregev** [comment]
> Can you make this part of an existing job?
> 
> These jobs are already flaky and each one added increases the chance for a PR to go red.

This PR is just meant to show the compilation problem. When it solve, we can think where to put the multissl.

File: lib/vtls/vtls.c
Comment (vszakats) [inline lib/vtls/vtls.c:424]: Is this an MSVC bug? Can we do a standard C initialization here to avoid silencing warnings?

Comment (talregev) [reply lib/vtls/vtls.c:424]: I don't think it msvc bug. But I cannot add standard C initialization because on the case where we don't have multi ssl, and it create a  multi redefinition, in line 1326

**Comment by talregev** [comment]
@vszakats I add another suggestion.

**Comment by talregev** [comment]
https://stackoverflow.com/questions/40738351/forward-declaration-as-struct-vs-class

FYI

File: lib/vtls/vtls.c
Comment (vszakats) [reply lib/vtls/vtls.c]: Have you tried moving the `Curl_ssl_cleanup` function below the initialization (e.g. before the function `multissl_version`), and dropping the declaration from line 417?

**Comment by talregev** [comment]
https://learn.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements-2017?view=msvc-170#uninitialized-const-variables

There is a problem with forward deceleration and const. Maybe it is msvc bug?

File: lib/vtls/vtls.c
Comment (talregev) [reply lib/vtls/vtls.c]: Yes. and it work. See my  new code.

**Comment by bagder** [comment]
> const object should be initialized

Global variables are guaranteed to be initialized to zero according to the standard. So what exactly does this mean?

**Comment by talregev** [comment]
> > const object should be initialized
> 
> Global variables are guaranteed to be initialized to zero according to the standard. So what exactly does this mean?

On the original code on line 417, it forward declaration, not a definition. It should not do initialization on this line.
For me it seems that you cannot do forward declaration with const in msvc, even if it legal in c.  That why I think it msvc bug.

**Comment by talregev** [comment]
https://learn.microsoft.com/th-th/cpp/cpp/declarations-and-definitions-cpp?view=msvc-170#definitions

**Comment by talregev** [comment]
https://stackoverflow.com/a/936589/550384

**Review by vszakats — CHANGES_REQUESTED** [review]
Looking good! As discussed earlier, I'd prefer integrating this into an existing job, rather than adding a new one to avoid adding more chance for flakiness.

How about enabling Schannel in the existing 'openssl' job?

**Comment by talregev** [comment]
I know you don't want to add many tests. but my plan is to have multissl to install as much ssl backends for curl together.
I already succeeded with vcpkg to have 4:  
```
C:\src\vcpkg\installed\x64-windows\tools\curl>curl --version
curl 8.9.0-DEV (Windows) libcurl/8.9.0-DEV GnuTLS/3.8.4 (mbedTLS/2.28.8) (OpenSSL/3.3.1) (Schannel) zlib/1.3.1
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe TLS-SRP Unicode UnixSockets
```

**Comment by vszakats** [comment]
Tests will only run on the default TLS backend, so running with extra backends will not increase test coverage. Having a single extra backend tests the MultiSSL codepath, this is covered in this PR. Adding more backends would only add _build tests_ for the extra backends. We have this covered for many now in separate jobs, and new ones can be appended to an existing job.

I think the situation with flaky MSVC jobs is already very bad, breaking almost all runs. I don't know why that it, but adding more flaky jobs don't seem to help us because of this.

All in all, I think we'd better off we using the existing `openssl` job to test MultiSSL, and untested backends can be appended there too.

**Comment by talregev** [comment]
Can we just build the tests without running them on multissl job?
If I add openssl to Schannel, openssl will be the default.

**Comment by vszakats** [comment]
Why would that be better than adding it to an existing job?

**Comment by talregev** [comment]
My goal is to add on the same build, as much backend together. If I add more backends in openssl test, it will be not be a default anymore, because curl choose the default automatic according the first in the alphabet order. 
As you can see this code is not compile with msvc on curl ci and then we found warning / errors on the code.
Also I want to find these error and warning, and try to fix them if possible.

**Comment by talregev** [comment]
I am not sure why the ci fail

**Comment by vszakats** [comment]
> My goal is to add on the same build, as much backend together. If I add more backends in openssl test, it will be not be a default anymore, because curl choose the default automatic according the first in the alphabet order. As you can see this code is not compile with msvc on curl ci and then we found warning / errors on the code. Also I want to find these error and warning, and try to fix them if possible.

You can manually select the default TLS backend using `-DCURL_DEFAULT_SSL_BACKEND=` if it becomes something else than OpenSSL.

The MultiSSL compiler warning came up because of enabling MultiSSL at all. Adding more backends will not uncover more such issues. Either way I've nowhere said not to enable more backends, but to add new to ones to an existing job.

Also notice that MultiSSL isn't really where the puck is going in curl, e.g. because it lacks support for HTTP/3 and the original reasons for having it are now obsolete.

**Comment by talregev** [comment]
ok. I will Enable Schannel in openssl job.
Thank you for your review, and time for listening my ideas.

**Comment by talregev** [comment]
we can wait until the PR #14305 will be merge, and I will rebase this PR from the master after the merge.

**Review by vszakats — DISMISSED** [review]

**Comment by vszakats** [comment]
We need to find another job for MultiSSL. The openssl one has HTTP/3 enabled which is incompatible with MultiSSL. Hence the CI failure.

How about the 'Schannel U' job? It needs the `-DCURL_DEFAULT_SSL_BACKEND=schannel` option for keeping Schannel the default.

**Comment by talregev** [comment]
Done.

**Review by vszakats — APPROVED** [review]

**Review by MarcelRaad — APPROVED** [review]

**Comment by vszakats** [comment]
Thanks @talregev!

**Pull request closed without merging**
